src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+              class = std::enable_if_t<IsAllowed>>
+    future(ErrorT&& e)
+      : base_t(
+          seastar::make_exception_future<ValuesT...>(
+            errorator<WrappedAllowedErrorsT...>::make_exception_ptr(e))) {
+      // this is `fold expression` of C++17. Because of the SFINAE above
+      // the assertion cannot fail. Keeping it just in the case.
+      static_assert((... || (std::is_same_v<DecayedT, WrappedAllowedErrorsT>)),
+                    "ErrorT is not enlisted in errorator");
+    }
+
+    template <class ValueFuncT, class ErrorVisitorT>
+    auto safe_then(ValueFuncT&& valfunc, ErrorVisitorT&& errfunc) {
+      static_assert((... && std::is_invocable_v<ErrorVisitorT,
+                                                WrappedAllowedErrorsT>),
+                    "provided Error Visitor is not exhaustive");
+
+      using ValueFuncResult = std::result_of_t<ValueFuncT(ValuesT&&...)>;
+      // recognize whether there can be any error coming from the Value
+      // Function.
+      using ValueFuncErrorator = get_errorator_t<ValueFuncResult>;
+      // mutate the Value Function's errorator to harvest errors coming
+      // from the Error Visitor. Yes, it's perfectly fine to fail error
+      // handling at one step and delegate even broader set of issues
+      // to next continuation.
+      using ReturnErrorator = make_errorator_t<
+        ValueFuncErrorator,
+        std::decay_t<std::result_of_t<
+          ErrorVisitorT(WrappedAllowedErrorsT)>>...>;
+      // OK, now we know about all errors next continuation must take
+      // care about. If Visitor handled everything and the Value Func
+      // doesn't return any, we'll finish with errorator<>::future
+      // which is just vanilla seastar::future – that's it, next cont
+      // finally could use `.then()`!
+      using Futurator = \
+        typename ReturnErrorator::template futurize<ValueFuncResult>;
+      // `seastar::futurize`, used internally by `then_wrapped()`, wraps
+      // any non-`seastar::future` in `seastar::future`. We really don't
+      // want to end with e.g. `seastar::future<errorator::future<...>>`,
+      // so the lambda converts errorated future into plain one and here
+      // we're converting it back. As there is absolutely no difference
+      // between these types from the memory layout's POV (casting could
+      // be used instead), I count on compiler's ability to elide copies.
+      return typename Futurator::type{ this->then_wrapped(
+        [ valfunc = std::forward<ValueFuncT>(valfunc),
+          errfunc = std::forward<ErrorVisitorT>(errfunc)
+        ] (auto future) mutable {
+          if (__builtin_expect(future.failed(), false)) {
+            maybe_handle_error_t<ErrorVisitorT, Futurator> maybe_handle_error(
+              std::forward<ErrorVisitorT>(errfunc),
+              std::move(future).get_exception()
+            );
+            (maybe_handle_error.template operator()<WrappedAllowedErrorsT>() , ...);
```

Comment:
Dumb question: how does this syntax work?  Specifically, I'm confused by the (<method call which returns void>, ...) syntax.

Reply:
That's fold expression using the comma (`,`) operator. It gets expanded to something like:

```cpp
call<T_1>(), call<T_2>(), ... , call<T_N>;
```

Reply:
I guess it's a tuple formed by applying operator() to each element of WrappedAllowedErrorsT in order to compactly apply the operator() method to each type.

Reply:
No tuple is involved – it's just expansion of the arg pack. The ultimate result of the expression is the value of last call which is actually also discarded – only side effects play role.

Reply:
Got it, thanks.